### PR TITLE
fix: sampler log regression

### LINKED ISSUE #3490: sampler stats regression
### Context
1.10.0 RC1 + updates

### Summary
The localstore sampler logs performance statistics when a sample is generated.   This stats entry used to contain useful counts and elapsed time durations, but some commit between 946fbc57 (readable) and 5f61be61 (not readable) broke the stats formatting.

### Expected behavior
I expected to be able to read the statistics logs.

### Actual behavior
Previous log (commit 946fbc57) looked like this (broken up for readability) - Note the stats entry:
```
"time"="2022-11-02 18:18:45.694569" "level"="info" "logger"="node/localstore/sampler" "msg"="Sampler done"
"Stats"="Total: 3740512 NotFound: 0 New Ignored: 1 Iteration Duration: 758 secs GetDuration: 4121 secs HmacrDuration: 90"
"Sample"={"Items":["000005382384f5df0b300281214f3413140dd499cf35bc95d083e481585ae3f4","000009a4e873493483e72441a838d4d2fad93eb6342fe2aafd0b78687e1dbf0a","00000e39721e3caf201ecfa8757299353b2976ba390b49b9d23e81780c513aaa","00000f5ba55d7b429c60bc9eb2fa4a79e4a37997d5d9cb1409313303c6cf3191","000014106a6851bad71fb03a4866ca3772f777fd5e1d569807d05a9bf4e5573e","000015f1de228d1cbb3b14eb12e93cfcf49d8019b72e090f2e12ab24daef41b0","00001602b52a72b92ad37d7b4d8e9fcc8405e4f1cfe6fb608d74c7bdf2b2bdbd","000020a4e7b984172f51d191a63224ca5e803bb95172472eaa3aa84a5f817599","000027d6459c4e1637e0b0a2cb29f0a8ee0364e0d7606b68472ccb3c37226152","000035c2790d1a1f9bca1297ce7fff00cf1729755314d819381ef4667144fae4","0000386f880d648db9ac845ab8ebd46ec7414318a1b4c1ff627aa8d480025827","0000392647a874677ffcb627cb48c4b2709b286b72bad0299d3ecf5be17e7221","00003a57fb5e35824578fb1f4b929f2c015f223cfa212037b95ff3b837b73602","00003d43dd047911ec05a4dd25514294c59c509e753618106cd37b261d037321","00003f608bfd228f1e272a561ec2893d620d7780f409b3a2be579d8207ca331c","000040323b23afbb8ee7f50b1f25b39bc99000d399618bd5f01ac27fc7046843"],"Hash":"43ebc209609c5622312532a35bff13e909ca5c860927759ae83d50e71b76e601"}
```

New log (commit 5f61be61) looks like this (broken up for readability) - Note the stats entry:
```
"time"="2022-11-02 19:26:50.996675" "level"="info" "logger"="node/localstore/sampler" "msg"="sampler done"
"stats"={"TotalIterated":{},"NotFound":{},"NewIgnored":{},"IterationDuration":{},"GetDuration":{},"HmacrDuration":{}}
"sample"={"Items":["000007670b56057cf551e724b4c02db8893c246d365981c62cd85a0fbe162e76","00000bb7d524daa7e58c891cfc400acb30388d3d4c835ab9ac27f6f4ef46b938","000010292cb6050ba755e552604e6a283b5dcd11ccad628872b88d699432a31b","000010453403479a6e97341670178ffa7e4e747772d3137d12e6c97d39254c79","00001cfcca8f83e0800eb5fed09c7138761b4f94b6f61b9afa1a618234b6efb3","00003398a253f7bb90edf2380b4b4e280351f76cff9bbd1310abc70def3f2a91","00003852c72c844d53b12b34d391d66d2bc6e2b4ec45e9ca2dee24e9d8d0841d","00003897d1e9dabbd336b9c0bd26ac45982e282cc9055012271722c90f09dfa1","0000391d58ee64a089107490aabbca0a83fba0b1a3092a5090f7e6b9aab7da34","00003af70e3526fa13b4bb9d32d45a23dceb052f384a489fbc3177d1811f7ad6","00003c07d647abab1ddc5bdd14cb5fac44261cf2a982ec9abb26a537e58931e0","00004a29f32057ebd3fcc1c9b007ee575c2f1cca775b29cc0c2f8e24d695cfed","000061eb18edc5c0b93f84e5d3412a26ae240c87a99ac59a076cfb89f6c78838","0000658cc5a70f2b7e06ec6313ecd6f1109181c380ef7ab4f6d88862725cef46","000066456ae64612140421a54aaa75ec93e6fe0789d2dd5b5fc3ff1a0e9abf06","000071048fa88431b9451f5b648a077de337c1f5eaa0a4d3d1618dd8532e3ed3"],"Hash":"266ef0fbeac6a1dfb4441a1e173225c53b33115767841f088b632abaadae3390"}
```

### Steps to reproduce
Just examine the info logs of any node that is actually generating a sample to commit and reveal to the storage incentives.

### Possible solution
Restore the original formatting for the stats item.

### PR DESCRIPTION
**Pull request by aloknerurkar**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [ ] I have added tests to cover my changes.
- [x] I have filled out the description and linked the related issues.

### Description
So seems like we need to pass the pointer to the stat item. This is because of how the logger internally handles printing the string.

Fixes #3490 

- - -
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3491)

### CODE REVIEW

**Review by istae — APPROVED** [review]

File: pkg/localstore/sampler.go [hunk 1]
```diff
@@ -232,7 +232,7 @@ func (db *DB) ReserveSample(
 		Items: sampleItems,
 		Hash:  swarm.NewAddress(hash),
 	}
-	logger.Info("sampler done", "stats", stat, "sample", sample)
+	logger.Info("sampler done", "stats", &stat, "sample", sample)
```
Comment (ldeffenb) [inline pkg/localstore/sampler.go:235]: Definitely tells me I've got a lot more to learn about golang!

**Review by ldeffenb — APPROVED** [review]
Definitely tells me I've got a lot more to learn about golang! But I'll take your word that it fixes the logging issue.

File: pkg/localstore/sampler.go (hunk 1 above)
Comment (mrekucci) [inline pkg/localstore/sampler.go:235]: Instead, I suggest changing the `String()` to a method without a pointer receiver.

**Review by mrekucci — APPROVED** [review]

**Comment by mrekucci** [comment]
> Definitely tells me I've got a lot more to learn about golang! But I'll take your word that it fixes the logging issue.

The simple answer can be summarised in the following points. If the receiver of your method is a pointer receiver, then this means that the type should be used as a pointer. Since the `sampleStat` method `String` had a pointer receiver but it was not used as a pointer, the type assertion mechanism was (is) unable to assert `String` method on a non-pointer type. This can be demonstrated in the following [example](https://go.dev/play/p/4LcnkvRU7ji).

**Pull request merged**